File: src/temporal/workflows/monthly_post.py
"""Monthly post workflow for trade confirmation bot."""

from datetime import timedelta

from temporalio import workflow

with workflow.unsafe.imports_passed_through():
    from .comment_processing import CommentPollingWorkflow

from bot.config import SUBREDDIT_NAME
from bot.models import CreateMonthlyPostInput, SubmissionInput

from ..activities import notifications as notification_activities
from ..activities import submissions as submission_activities
from ..shared import PUSHOVER_RETRY_POLICY
from ..shared import REDDIT_RETRY_POLICY_CONSERVATIVE as REDDIT_RETRY_POLICY


@workflow.defn
class MonthlyPostWorkflow:
    """Creates monthly confirmation thread and locks the previous one after a delay.

    This workflow is scheduled to run on the 1st of each month.
    It handles:
    - Querying the polling workflow for the current submission ID
    - Creating the new monthly post
    - Signalling the polling workflow immediately with the new submission
    - Stickying the new post and unstickying the old one
    - Sleeping for 5 days, then locking the old submission
    """

    @workflow.run
    async def run(self) -> dict[str, str | None]:
        """Create the monthly confirmation thread and lock the old one after 5 days.

        Returns:
            Result with status and submission_id.
        """
        workflow.logger.info("Starting monthly post workflow for r/%s", SUBREDDIT_NAME)

        await workflow.execute_activity(
            notification_activities.send_pushover_notification,
            args=[f"Creating monthly post for r/{SUBREDDIT_NAME}"],
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=PUSHOVER_RETRY_POLICY,
        )

        # Discover the current submission ID (the one we're replacing).
        active_submissions = await workflow.execute_activity(
            submission_activities.fetch_active_submission_ids,
            start_to_close_timeout=timedelta(seconds=60),
            retry_policy=REDDIT_RETRY_POLICY,
        )
        old_submission_id = active_submissions.current_submission_id

        # Create the new monthly post (idempotent).
        new_submission_id = await workflow.execute_activity(
            submission_activities.create_monthly_post,
            args=[CreateMonthlyPostInput(previous_submission_id=old_submission_id)],
            start_to_close_timeout=timedelta(seconds=60),
            retry_policy=REDDIT_RETRY_POLICY,
        )

        # Signal the polling workflow immediately so it starts scanning the
        # new submission without waiting for sticky/unsticky to finish.
        try:
            polling_handle = workflow.get_external_workflow_handle(
                f"poll-{SUBREDDIT_NAME}"
            )
            await polling_handle.signal(CommentPollingWorkflow.set_current_submission, new_submission_id)
            workflow.logger.info("Signalled polling workflow with new submission")
        except Exception as exc:
            workflow.logger.warning(
                "Could not signal polling workflow: %s: %s",
                type(exc).__name__,
                exc,
            )

        # Sticky the new post, unsticky the old one.
        await workflow.execute_activity(
            submission_activities.sticky_submission,
            args=[SubmissionInput(submission_id=new_submission_id)],
            start_to_close_timeout=timedelta(seconds=60),
            retry_policy=REDDIT_RETRY_POLICY,
        )

        if old_submission_id:
            await workflow.execute_activity(
                submission_activities.unsticky_submission,
                args=[SubmissionInput(submission_id=old_submission_id)],
                start_to_close_timeout=timedelta(seconds=60),
                retry_policy=REDDIT_RETRY_POLICY,
            )

        await workflow.execute_activity(
            notification_activities.send_pushover_notification,
            args=[f"Monthly post for r/{SUBREDDIT_NAME}: {new_submission_id}"],
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=PUSHOVER_RETRY_POLICY,
        )

        workflow.logger.info("Monthly post created: %s", new_submission_id)

        # Wait 5 days then lock the old submission.
        if old_submission_id:
            await workflow.sleep(timedelta(days=5))

            await workflow.execute_activity(
                submission_activities.lock_submission,
                args=[SubmissionInput(submission_id=old_submission_id)],
                start_to_close_timeout=timedelta(seconds=60),
                retry_policy=REDDIT_RETRY_POLICY,
            )

            await workflow.execute_activity(
                notification_activities.send_pushover_notification,
                args=[
                    f"Locked previous submission {old_submission_id} for r/{SUBREDDIT_NAME}"
                ],
                start_to_close_timeout=timedelta(seconds=30),
                retry_policy=PUSHOVER_RETRY_POLICY,
            )

            workflow.logger.info("Locked previous submission: %s", old_submission_id)

        return {
            "status": "created",
            "submission_id": new_submission_id,
            "locked_submission_id": old_submission_id,
        }

File: internal/workflows/monthly_post.go
package workflows

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"

	"github.com/mikeacjones/reddit-trade-confirmation-bot/internal/models"
	"github.com/mikeacjones/reddit-trade-confirmation-bot/internal/shared"
)

// MonthlyPostWorkflow creates the monthly confirmation thread and locks the old one after 5 days.
func MonthlyPostWorkflow(ctx workflow.Context) (models.MonthlyPostResult, error) {
	logger := workflow.GetLogger(ctx)
	logger.Info("Starting monthly post workflow", "subreddit", SubredditName)

	notify := func(msg, summary string) {
		ao := workflow.ActivityOptions{
			StartToCloseTimeout: 30 * time.Second,
			RetryPolicy:         shared.PushoverRetry,
			Summary:             summary,
		}
		_ = workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, ao), "send_pushover_notification", msg).Get(ctx, nil)
	}

	notify(fmt.Sprintf("Creating monthly post for r/%s", SubredditName), "monthly-start:r/"+SubredditName)

	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 60 * time.Second,
		RetryPolicy:         shared.RedditRetryConservative,
		Summary:             "r/" + SubredditName,
	}
	var active models.ActiveSubmissions
	if err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, ao), "fetch_active_submission_ids").Get(ctx, &active); err != nil {
		return models.MonthlyPostResult{}, err
	}
	oldSubmissionID := active.CurrentSubmissionID

	summary := "prev:none"
	if oldSubmissionID != nil {
		summary = "prev:" + *oldSubmissionID
	}
	cao := workflow.ActivityOptions{
		StartToCloseTimeout: 60 * time.Second,
		RetryPolicy:         shared.RedditRetryConservative,
		Summary:             summary,
	}
	var newSubmissionID string
	if err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, cao), "create_monthly_post", models.CreateMonthlyPostInput{
		PreviousSubmissionID: oldSubmissionID,
	}).Get(ctx, &newSubmissionID); err != nil {
		return models.MonthlyPostResult{}, err
	}

	// Signal polling workflow
	pollingID := "poll-" + SubredditName
	if err := workflow.SignalExternalWorkflow(ctx, pollingID, "", "set_current_submission", newSubmissionID).Get(ctx, nil); err != nil {
		logger.Warn("Could not signal polling workflow", "error", err)
	} else {
		logger.Info("Signalled polling workflow with new submission")
	}

	sao := workflow.ActivityOptions{
		StartToCloseTimeout: 60 * time.Second,
		RetryPolicy:         shared.RedditRetryConservative,
		Summary:             newSubmissionID,
	}
	if err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, sao), "sticky_submission", newSubmissionID).Get(ctx, nil); err != nil {
		return models.MonthlyPostResult{}, err
	}

	if oldSubmissionID != nil {
		uao := workflow.ActivityOptions{
			StartToCloseTimeout: 60 * time.Second,
			RetryPolicy:         shared.RedditRetryConservative,
			Summary:             *oldSubmissionID,
		}
		if err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, uao), "unsticky_submission", *oldSubmissionID).Get(ctx, nil); err != nil {
			return models.MonthlyPostResult{}, err
		}
	}

	notify(fmt.Sprintf("Monthly post for r/%s: %s", SubredditName, newSubmissionID), "monthly-post:"+newSubmissionID)
	logger.Info("Monthly post created", "id", newSubmissionID)

	if oldSubmissionID != nil {
		_ = workflow.Sleep(ctx, 5*24*time.Hour)

		lao := workflow.ActivityOptions{
			StartToCloseTimeout: 60 * time.Second,
			RetryPolicy:         shared.RedditRetryConservative,
			Summary:             *oldSubmissionID,
		}
		if err := workflow.ExecuteActivity(workflow.WithActivityOptions(ctx, lao), "lock_submission", *oldSubmissionID).Get(ctx, nil); err != nil {
			return models.MonthlyPostResult{}, err
		}
		notify(
			fmt.Sprintf("Locked previous submission %s for r/%s", *oldSubmissionID, SubredditName),
			"locked:"+*oldSubmissionID,
		)
		logger.Info("Locked previous submission", "id", *oldSubmissionID)
	}

	return models.MonthlyPostResult{
		Status:             "created",
		SubmissionID:       newSubmissionID,
		LockedSubmissionID: oldSubmissionID,
	}, nil
}
